modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/components/rule-builder/Calculator/CalculatorButtonArea.es.js

```diff
@@ -0,0 +1,229 @@
+					className="border-top-left calculator-button"
+					displayType="secondary"
+					onClick={() =>
+						onCalculatorInput({
+							tokenType: TOKEN_TYPES.BACKSPACE,
+							tokenValue: 'backspace',
+						})
+					}
+					symbol="angle-left"
+					title={Liferay.Language.get('backspace')}
+				/>
+				<ClayButton
+					aria-label={Liferay.Language.get('left-parenthesis')}
+					className="calculator-button"
+					disabled={disableNumbers}
+					displayType="secondary"
+					monospaced
+					onClick={() =>
+						onCalculatorInput({
+							tokenType: TOKEN_TYPES.PARENTHESIS_LEFT,
+							tokenValue: '(',
+						})
+					}
+					title={Liferay.Language.get('left-parenthesis')}
+				>
+					{'('}
+				</ClayButton>
+
+				<ClayButton
+					aria-label={Liferay.Language.get('right-parenthesis')}
+					className="border-top-right calculator-button"
+					disabled={disableNumbers}
+					displayType="secondary"
+					monospaced
+					onClick={() =>
+						onCalculatorInput({
+							tokenType: TOKEN_TYPES.PARENTHESIS_RIGHT,
+							tokenValue: ')',
+						})
+					}
+					title={Liferay.Language.get('right-parenthesis')}
+				>
+					{')'}
+				</ClayButton>
+				{ONE_TO_NINE.map((value) => (
+					<ClayButton
+						aria-label={value}
+						className="calculator-button"
+						disabled={disableNumbers}
+						displayType="secondary"
+						key={value}
+						monospaced
+						onClick={() =>
+							onCalculatorInput({
+								tokenType: TOKEN_TYPES.LITERAL,
+								tokenValue: value,
+							})
+						}
+					>
+						{value}
+					</ClayButton>
+				))}
+				<ClayButton
+					aria-label={Liferay.Language.get('zero')}
+					className="border-bottom-left button-two-columns calculator-button"
+					disabled={disableNumbers}
+					displayType="secondary"
+					monospaced
+					onClick={() =>
+						onCalculatorInput({
+							tokenType: TOKEN_TYPES.LITERAL,
+							tokenValue: '0',
+						})
+					}
+				>
+					{'0'}
+				</ClayButton>
+				<ClayButton
+					aria-label={Liferay.Language.get('period')}
+					className="border-bottom-right calculator-button"
+					disabled={disableDot || disableNumbers}
+					displayType="secondary"
+					monospaced
+					onClick={() =>
+						onCalculatorInput({
+							tokenType: TOKEN_TYPES.LITERAL,
+							tokenValue: '.',
+						})
+					}
+				>
+					{'.'}
+				</ClayButton>
+			</div>
+
+			<div className="calculator-buttons calculator-buttons-operators">
+				<div className="btn-group-vertical" role="group">
+					<Dropdown
+						className="dropdown-action"
+						items={functions}
+						onItemSelected={onFunctionSelected}
+						resultSelected={resultSelected}
+						trigger={
+							<ClayButtonWithIcon
+								aria-label={Liferay.Language.get('actions')}
+								className="calculator-add-operator-button-area"
+								disabled={disableFunctions}
+								symbol="ellipsis-v"
+								title={Liferay.Language.get('actions')}
+							/>
+						}
+					/>
```

Comment:
mmmmm, I can try adding the callbacks for each functions but We will have two problems here:

1. depending the number of functions the customers create, We gonna have at least O(n) for just formatting those things before passing to the dropdown.

2. introduce a bad pattern on the code

once https://github.com/liferay/clay/issues/3776 is solved it will reduce the payload of transformations here.

Reply:
I'm not so sure this would be a problem O(n), this list seems to have a fixed size on the backend side, I may be mistaken but I just see a function being added, and I find it very unlikely to have a list greater than 100, it would still be a small number. So don't worry about that, JavaScript can handle a bigger list compared to this using `map`.

Otherwise, we will have to deal with the rendering of these elements with a virtualized list.

> introduce a bad pattern on the code

Why? The data was written for the old structure so we only need to organize for the new structure if possible. I don't see that that would be a bad practice.

```js
<ClayDropDownWithItems
	items={items.map((item) => ({...item, onClick}))}
	trigger={
		<>...</>
	}
/>
```

It is not a bad practice to transform the data.

Reply:
Hey Matu, it's simple to replace [Calculator.es.js usage](https://github.com/liferay-forms/liferay-portal/pull/28/files#diff-8014fad73b90f8aaea4a3db7b8abe2e49d71d57fb368338425b60e700c92dc29R222) but for [CalculatorButtonArea component we have a customization for the Dropdown.Item](https://github.com/liferay-forms/liferay-portal/pull/28/files#diff-40887227d1ae73e2f144a752c6cc6b42ed426b139d75b8b64dbb101bc6e6e94eR148).

Reply:
When seeing Dropdown component, I saw a missing condition there:
![image](https://user-images.githubusercontent.com/7663513/98740094-416baf00-2389-11eb-895e-d298f14e818e.png)

Reply:
@diegonvs then we can move the Dropdown code to CalculatorButtonArea and keep the high-level where we can use it. Just to avoid having a custom DropDown component.

Reply:
Done papi!